### PR TITLE
Handle reply-toggling for headless threads

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR is a review-ready variant of https://github.com/hypothesis/client/pull/2932 — please see that PR for (long) description.

New working branch here because of a complex rebasing situation and a refactoring of some components. Changes since https://github.com/hypothesis/client/pull/2932:

* `AnnotationReplyToggle` is now a "dumb" component per suggestion.
* `Annotation` and `AnnotationMissing` props are more "harmonized" now; `AnnotationMissing` props are a subset of `Annotation` props. Minor refactoring of `Thread` and `Annotation` to accommodate this.
* Tests added for everything here.

Fixes https://github.com/hypothesis/client/issues/2865

### CODE REVIEW

**Review by esanzgar — APPROVED** [review]
Great PR!

File: src/sidebar/components/AnnotationMissing.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import classnames from 'classnames';
+import { createElement } from 'preact';
+import propTypes from 'prop-types';
+
+import AnnotationReplyToggle from './AnnotationReplyToggle';
+
+/** @typedef {import('./Annotation').AnnotationProps} AnnotationProps */
+
+/**
+ * @typedef {Omit<AnnotationProps, "annotation" | "showDocumentInfo" | "annotationsService">} AnnotationMissingProps
```
Comment (esanzgar) [inline src/sidebar/components/AnnotationMissing.js:10]: This is a good idea, to mimic the props definition.

File: src/sidebar/components/AnnotationReplyToggle.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { createElement } from 'preact';
+import propTypes from 'prop-types';
+
+import Button from './Button';
+
+/**
+ * @typedef AnnotationReplyToggleProps
+ * @prop {() => any} onToggleReplies
```
[the remark this answers is not in the document]
Comment (lyzadanger) [reply src/sidebar/components/AnnotationReplyToggle.js:8]: Oooh, I am a newly-minted fan of `VoidFunction`. So much easier to read, IMHO. Thanks!

File: src/sidebar/components/AnnotationMissing.js (hunk 1 above)
Comment (lyzadanger) [reply src/sidebar/components/AnnotationMissing.js:10]: Yeah, it felt kind of right... I just made a small change here to use single quotes and to not have spaces between keys in the union (per some feedback on another PR). I'll bring up typedef style conventions at our next frontend sync, as I feel like we might be getting divergent.

**Pull request merged**